### PR TITLE
FEAT: BAC-9 Forgot Password and Password Reset Website

### PR DESCRIPTION
**Pull request by Mofe620**
- Verifies if email provided by user is registered on our site
- Added endpoint and functionality to send password reset link to verified user email 
- Added secret token to reset link in order to protect the update-password route
- Stored the hash of the secret token in a database `tokens` table
- Requires token from reset link to be verified with the hash stored in the database before updating password
- Added update password functionality

Linear Ticket:
- https://linear.app/team-opspad/issue/BAC-9/forgot-password-and-password-reset-website [5 points]

_Mofentan Afinotan_

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -20,18 +20,21 @@
   "dependencies": {
     "bcrypt": "^5.1.0",
     "cors": "^2.8.5",
+    "crypto": "^1.0.1",
```
Comment (orlando70) [inline backend/package.json:23]: You didn't need to install this package. It now comes with nodeJs.

File: backend/src/database/migrations/20221201094044_create-token.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export async function up(knex) {
+    return knex.schema.createTable("tokens", (table) => {
+        table.increments("id").primary();
```
Comment (orlando70) [inline backend/src/database/migrations/20221201094044_create-token.js:3]: We are using UUID not incremental id's. Make this change.

File: backend/src/controllers/AuthController.js [hunk 3]
```diff
@@ -1,30 +1,84 @@
...
+            res.status(200).redirect("/login");
+            return ({message: "A password reset link has been sent to your email address"});
+        
+        } catch (error) {
+            console.error(error);
+        }
+    };
+
+    static updateUserPassword = async (req, res) => {
+        const {token, id, password} = req.body;
```
Comment (orlando70) [inline backend/src/controllers/AuthController.js:57]: You sent these out as request params, now you're expecting it from request body.

File: backend/src/controllers/AuthController.js [hunk 4]
```diff
@@ -1,30 +1,84 @@
...
+        if(!updatedUser)throw new NotFoundError("User not found");
+
+        res.status(200).redirect("/login");
+        return ({   
+            success: true,
+            message: "Password has been updated successfully",
+            updatedUser
+        });
+    };
+}
```
Comment (orlando70) [inline backend/src/controllers/AuthController.js:84]: I don't see where you delete this token after it has achieved it's purpose. This means someone could reuse his/her tokens infinitely. Please delete the tokens after it has served it's purpose.

**Review by orlando70 — CHANGES_REQUESTED** [review]

File: backend/src/controllers/AuthController.js (hunk 3 above)
Comment (Mofe620) [reply backend/src/controllers/AuthController.js:57]: No. Not really. The frontend guy retrieves them from the url and puts them into the payload alongside the new password to be set. I've already explained that to @dread557

File: backend/package.json (hunk 1 above)
Comment (Mofe620) [reply backend/package.json:23]: So I should uninstall it yeah?

File: backend/src/database/migrations/20221201094044_create-token.js (hunk 2 above)
Comment (Mofe620) [reply backend/src/database/migrations/20221201094044_create-token.js]: Fixed.

File: backend/package.json (hunk 1 above)
Comment (Mofe620) [reply backend/package.json]: Fixed

File: backend/src/controllers/AuthController.js (hunk 4 above)
Comment (Mofe620) [reply backend/src/controllers/AuthController.js:88]: Fixed

**Review by orlando70 — APPROVED** [review]

**Review by orlando70 — APPROVED** [review]

**Review by orlando70 — APPROVED** [review]

**Pull request merged**